clippy_lints/src/manual_checked_sub.rs

```diff
@@ -0,0 +1,239 @@
+use clippy_config::Conf;
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::snippet_with_applicability;
+use rustc_ast::{BinOpKind, LitIntType, LitKind};
+use rustc_data_structures::packed::Pu128;
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::intravisit::{Visitor, walk_expr};
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self};
+use rustc_session::impl_lint_pass;
+use std::fmt::Write;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for manual re-implementations of checked subtraction.
+    ///
+    /// ### Why is this bad?
+    /// Manually re-implementing checked subtraction can be error-prone and less readable.
+    /// Using the standard library method `.checked_sub()` is clearer and less likely to contain bugs.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// if 10u32 >= 4u32 {
+    ///     let c = 10u32 - 4u32;
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// if let Some(result) = 10u32.checked_sub(4u32){
+    ///    result;
+    /// };
+    /// ```
+    #[clippy::version = "1.86.0"]
+    pub MANUAL_CHECKED_SUB,
+    complexity,
+    "Checks for manual re-implementations of checked subtraction."
+}
+
+pub struct ManualCheckedSub {
+    msrv: Msrv,
+}
+
+impl ManualCheckedSub {
+    #[must_use]
+    pub fn new(conf: &'static Conf) -> Self {
+        Self { msrv: conf.msrv }
+    }
+}
+
+impl_lint_pass!(ManualCheckedSub => [MANUAL_CHECKED_SUB]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualCheckedSub {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if !self.msrv.meets(cx, msrvs::MANUAL_CHECKED_SUB) {
+            return;
+        }
+
+        let applicability = Applicability::MachineApplicable;
+
+        if let ExprKind::If(drop_temp_expr, body_block, else_block) = expr.kind
+            && let ExprKind::DropTemps(condition_expr) = drop_temp_expr.kind
```

Comment:
By using `clippy_utils::higher::If::hir(expr)`, you would not have to deal with `DropTemps`, especially if at one point the HIR changes, that would be covered by the `higher` module for every lint.